README.md
# CGoggles

[![Code Style](https://img.shields.io/badge/code_style-VS_Code-blue.svg?style=flat)](https://google.github.io/styleguide/cppguide.html)
[![Versioning](https://img.shields.io/badge/versioning-semantic-brightgreen.svg?style=flat)](https://semver.org/)
[![Commit Style](https://img.shields.io/badge/commit_style-gitmoji-yellow.svg?style=flat)](https://gitmoji.carloscuesta.me/)
![Top Language](https://img.shields.io/github/languages/top/evaneliasyoung/cgoggles.svg?style=flat)
![Code Size](https://img.shields.io/github/languages/code-size/evaneliasyoung/cgoggles.svg?style=flat)
[![Issues](https://img.shields.io/github/issues/evaneliasyoung/cgoggles.svg?style=flat)](https://github.com/evaneliasyoung/cgoggles/issues)
[![Last Commit](https://img.shields.io/github/last-commit/evaneliasyoung/cgoggles.svg?style=flat)](https://github.com/evaneliasyoung/cgoggles/commit/master)

The ultimate cross-platform System information CLI tool.
Written in C++ to allow for speedy-compile times and speedy-operation times.
This is made as first a passion-project and next a final-project for my programming class.

## News and Changes

### Version 0.6.0 - 2020-02-29

### Latest Activity

(last 7 major and minor version releases)

- Version 0.5.0: Catalina support for macOS, Linux bug fixes, performance improvements
- Version 0.4.0: Full Linux support
- Version 0.3.3: Doxygen documentation, performance improvements, value-only output
- Version 0.3.2: Performance improvements, RAM information for Mac
- Version 0.3.1: Pretty-print option for storage
- Version 0.3.0: System information and most Linux support
- Version 0.2.0: GPU, Storage, and FileSystem information

You can find all changes here: [detailed changelog](CHANGELOG.md)

## Reference

~ Partial Support

X Complete Support

S Sudo Support


### 1. OS

| Call           | Linux | Mac | Win | Comments                 |
| -------------- | ----- | --- | --- | ------------------------ |
| os             | ~     | X   | X   | *Everything below*       |
| os.All         | ~     | X   | X   | *Everything below*       |
| os.Platform    | X     | X   | X   | Windows                  |
| os.Caption     | X     | X   | X   | Microsoft Windows 10 Pro |
| os.Serial      | S     | X   | X   | 00000-00000-00000-XXXXX  |
| os.Bit         | X     | X   | X   | 64                       |
| os.InstallTime | X     | X   | X   | 2018-12-04T04:41:06      |
| os.BootTime    | X     | X   | X   | 2019-03-22T18:03:37      |
| os.CurTime     | X     | X   | X   | 2019-03-25T17:49:18      |
| os.Kernel      | X     | X   | X   | 10.0.0                   |
| os.Version     | X     | X   | X   | 10.0.0                   |

### 2. System
| Call             | Linux | Mac | Win | Comments                             |
| ---------------- | ----- | --- | --- | ------------------------------------ |
| sys              | X     | X   | X   | *Everything below*                   |
| sys.All          | X     | X   | X   | *Everything below*                   |
| sys.Manufacturer | X     | X   | X   | Apple Inc.                           |
| sys.Model        | X     | X   | X   | MacBookPro11,5                       |
| sys.Version      | X     | X   | X   | 1.0                                  |
| sys.Serial       | S     | X   | X   | XXXXXXXXXXXX                         |
| sys.UUID         | S     | X   | X   | XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX |

### 2. CPU

| Call             | Linux | Mac | Win | Comments                             |
| ---------------- | ----- | --- | --- | ------------------------------------ |
| cpu              | X     | X   | X   | *Everything below*                   |
| cpu.All          | X     | X   | X   | *Everything below*                   |
| cpu.Manufacturer | X     | X   | X   | AuthenticAMD                         |
| cpu.Architecture | X     | X   | X   | x64                                  |
| cpu.SocketType   | X     | X   | X   | AM4                                  |
| cpu.Brand        | X     | X   | X   | AMD Ryzen 5 1600x Six-Core Processor |
| cpu.Family       | X     | X   | X   | 23                                   |
| cpu.Model        | X     | X   | X   | 1                                    |
| cpu.Stepping     | X     | X   | X   | 1                                    |
| cpu.Cores        | X     | X   | X   | 6                                    |
| cpu.Threads      | X     | X   | X   | 12                                   |
| cpu.Speed        | X     | X   | X   | 3.620 GHz                            |
| cpu.MaxSpeed     | X     | X   | X   | 3.620 GHz                            |

### 3. GPU

| Call        | Linux | Mac | Win | Comments           |
| ----------- | ----- | --- | --- | ------------------ |
| gpu         |       | X   | X   | *Everything below* |
| gpu.All     |       | X   | X   | *Everything below* |
| gpu.Vendor  |       | X   | X   | Intel              |
| gpu.Model   |       | X   | X   | Intel Iris Pro     |
| gpu.Bus     |       | X   | X   | Built-In           |
| gpu.VRAM    |       | X   | X   | 1536               |
| gpu.Dynamic |       | X   | X   | Yes                |

### 4. RAM

| Call                  | Linux | Mac | Win | Comments           |
| --------------------- | ----- | --- | --- | ------------------ |
| ram                   | S     | ~   | ~   | *Everything below* |
| ram.All               | S     | ~   | ~   | *Everything below* |
| ram.Size              | S     | X   | X   | 8589934592         |
| ram.Bank              | S     | X   | X   | DIMM1              |
| ram.Type              | S     | X   | X   | DDR4               |
| ram.Speed             | S     | X   | X   | 3200 MHz           |
| ram.FormFactor        | S     |     | X   | DIMM               |
| ram.Manufacturer      | S     | X   | X   | Samsung            |
| ram.Part              | S     | X   | X   | M378B1G73EB0-YK0   |
| ram.Serial            | S     | X   | X   | XXXXXXXXXXXX       |
| ram.VoltageConfigured | S     |     | X   | 2.4 V              |
| ram.VoltageMin        | S     |     | X   | 2.4 V              |
| ram.VoltageMax        | S     |     | X   | 2.4 V              |

### 5. Storage

| Call               | Linux | Mac | Win | Comments                             |
| ------------------ | ----- | --- | --- | ------------------------------------ |
| storage            | X     | ~   | ~   | *Everything below*                   |
| storage.All        | X     | ~   | ~   | *Everything below*                   |
| storage.Name       | X     | X   | X   | Name                                 |
| storage.Type       | X     | X   | X   | Type                                 |
| storage.FileSystem | X     | X   | X   | NTFS                                 |
| storage.Mount      | X     | X   | X   | /dev/disk0s1                         |
| storage.Size       | X     | X   | X   | 500277790720                         |
| storage.Physical   | X     | X   | X   | 5002HDD77790720                      |
| storage.UUID       | X     | X   | X   | XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX |
| storage.Label      | X     | X   | X   | VM                                   |
| storage.Model      | X     | X   |     | APPLE SSD SM0512G                    |
| storage.Serial     | X     |     | X   | XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX |
| storage.Removable  | X     | X   | X   | No                                   |
| storage.Protocol   | X     | X   |     | PCI                                  |

### 6. File Systems

| Call     | Linux | Mac | Win | Comments             |
| -------- | ----- | --- | --- | -------------------- |
| fs       | X     | X   | X   | *Everything below*   |
| fs.All   | X     | X   | X   | *Everything below*   |
| fs.FS    | X     | X   | X   | /dev/disk1s1         |
| fs.Type  | X     | X   | X   | HFS                  |
| fs.Size  | X     | X   | X   | 18446744072211382272 |
| fs.Used  | X     | X   | X   | 18446744073346375680 |
| fs.Mount | X     | X   | X   | /                    |


CHANGELOG.md
# Changelog

## Version History

| Version | Date       | Comment                                                               |
| ------- | ---------- | --------------------------------------------------------------------- |
| 0.5.0   | 2020-02-29 | Catalina support for macOS, Linux bug fixes, performance improvements |
| 0.4.0   | 2019-04-30 | Full Linux support                                                    |
| 0.3.3   | 2019-04-29 | Doxygen documentation, performance improvements, value-only output    |
| 0.3.2   | 2019-04-22 | Performance improvements, RAM information for Mac                     |
| 0.3.1   | 2019-04-17 | Pretty-print option for storage                                       |
| 0.3.0   | 2019-04-12 | System information and most Linux support                             |
| 0.2.0   | 2019-04-01 | GPU, Storage, and FileSystem information                              |
| 0.1.1   | 2019-03-28 | Mac compatability                                                     |
| 0.1.0   | 2019-03-25 | Initial release                                                       |


os.cpp
/**
*  @file      os.cpp
*  @brief     The implementation for the OperatingSystem Class.
*
*  @author    Evan Elias Young
*  @date      2019-03-15
*  @date      2020-02-29
*  @copyright Copyright 2019-2020 Evan Elias Young. All rights reserved.
*/

#include "pch.h"
#include "os.h"
#include "semver.h"
#include "utils.h"

#pragma region "Constructors"
/**
* @brief Construct a new OperatingSystem object
*/
OperatingSystem::OperatingSystem()
{
  platform = "";
  caption = "";
  serial = "";
  bit = 0;
  installTime = std::tm();
  bootTime = std::tm();
  curTime = std::tm();
  kernel = SemVer();
  version = SemVer();
}

/**
* @brief Construct a new OperatingSystem object with help from the assistants
*
* @param plt The platform of the system
*/
OperatingSystem::OperatingSystem(std::uint8_t plt)
{
  platform = "";
  caption = "";
  serial = "";
  bit = 0;
  installTime = std::tm();
  bootTime = std::tm();
  curTime = std::tm();
  kernel = SemVer();
  version = SemVer();

  switch (plt)
  {
  case OS_WIN:
    GetWin();
    break;
  case OS_MAC:
    GetMac();
    break;
  case OS_LUX:
    GetLux();
    break;
  }
}

/**
* @brief Construct a new OperatingSystem object from another OperatingSystem object
*
* @param o The OperatingSystem object to copy from
*/
OperatingSystem::OperatingSystem(const OperatingSystem &o)
{
  platform = o.platform;
  caption = o.caption;
  serial = o.serial;
  bit = o.bit;
  installTime = o.installTime;
  bootTime = o.bootTime;
  curTime = o.curTime;
  kernel = o.kernel;
  version = o.version;
}
#pragma endregion "Constructors"

#pragma region "Constructors' Assistants"
/**
* @brief Fills in the OS information for Darwin systems
*/
void OperatingSystem::GetMac()
{
  std::string versionCapts[14] = {
      "Jaguar",
      "Panther",
      "Tiger",
      "Leopard",
      "Snow Leopard",
      "Lion",
      "Mountain Lion",
      "Mavericks",
      "Yosemite",
      "El Capitan",
      "Sierra",
      "High Sierra",
      "Mojave",
      "Catalina"};
  std::string temp;
  std::time_t tempTime;

  platform = "Darwin";
  serial = runCommand("system_profiler SPHardwareDataType | grep Serial | awk '{ print $4; }'");
  trim(&serial);
  bit = std::stoi(runCommand("getconf LONG_BIT"));
  version = new SemVer(runCommand("sysctl -n kern.osproductversion"), 0b11100u);
  kernel = new SemVer(runCommand("sysctl -n kern.osrelease"), 0b11100u);
  caption = version.Minor() > 11 ? "macOS" : "Mac OS X";
  if (version.Major() == 10 && version.Minor() > 1 && version.Minor() < 15)
  {
    caption += " " + versionCapts[version.Minor() - 2];
  }

  temp = runCommand(R"(system_profiler SPInstallHistoryDataType | grep -A 4 -E '^\s{4}(macOS|OS X)' | tail -n1 | awk '{ print $3 $4; }')");
  installTime.tm_year = std::stoi(temp.substr(0, 4)) - 1900;
  installTime.tm_mon = std::stoi(temp.substr(5, 2)) - 1;
  installTime.tm_mday = std::stoi(temp.substr(8, 2));
  installTime.tm_hour = std::stoi(temp.substr(11, 2));
  installTime.tm_min = std::stoi(temp.substr(14, 2));

  tempTime = (std::time_t)std::stoi(runCommand("sysctl kern.boottime | awk '{ print substr($5, 1, length($5) - 1); }'"));
  bootTime = (*std::localtime(&tempTime));

  temp = runCommand(R"(date +"%Y%m%d%H%M%S")");
  curTime.tm_year = std::stoi(temp.substr(0, 4)) - 1900;
  curTime.tm_mon = std::stoi(temp.substr(4, 2)) - 1;
  curTime.tm_mday = std::stoi(temp.substr(6, 2));
  curTime.tm_hour = std::stoi(temp.substr(8, 2));
  curTime.tm_min = std::stoi(temp.substr(10, 2));
  curTime.tm_sec = std::stoi(temp.substr(12, 2));
}

/**
* @brief Fills in the OS information for Windows systems
*/
void OperatingSystem::GetWin()
{
  std::string wmic = getWmicPath();
  std::map<std::string, std::string> dataMap = runMultiWmic("os get Caption,SerialNumber,OSArchitecture,Version,Version,InstallDate,LastBootUpTime,LocalDateTime", &wmic);
  std::string temp;

  platform = "Windows";
  caption = dataMap["Caption"];
  serial = dataMap["SerialNumber"];
  bit = std::stoi(dataMap["OSArchitecture"].erase(3));
  version = new SemVer(dataMap["Version"], 0b11010u);
  kernel = new SemVer(dataMap["Version"], 0b11010u);

  temp = dataMap["InstallDate"];
  installTime.tm_year = std::stoi(temp.substr(0, 4)) - 1900;
  installTime.tm_mon = std::stoi(temp.substr(4, 2)) - 1;
  installTime.tm_mday = std::stoi(temp.substr(6, 2));
  installTime.tm_hour = std::stoi(temp.substr(8, 2));
  installTime.tm_min = std::stoi(temp.substr(10, 2));
  installTime.tm_sec = std::stoi(temp.substr(12, 2));

  temp = dataMap["LastBootUpTime"];
  bootTime.tm_year = std::stoi(temp.substr(0, 4)) - 1900;
  bootTime.tm_mon = std::stoi(temp.substr(4, 2)) - 1;
  bootTime.tm_mday = std::stoi(temp.substr(6, 2));
  bootTime.tm_hour = std::stoi(temp.substr(8, 2));
  bootTime.tm_min = std::stoi(temp.substr(10, 2));
  bootTime.tm_sec = std::stoi(temp.substr(12, 2));

  temp = dataMap["LocalDateTime"];
  curTime.tm_year = std::stoi(temp.substr(0, 4)) - 1900;
  curTime.tm_mon = std::stoi(temp.substr(4, 2)) - 1;
  curTime.tm_mday = std::stoi(temp.substr(6, 2));
  curTime.tm_hour = std::stoi(temp.substr(8, 2));
  curTime.tm_min = std::stoi(temp.substr(10, 2));
  curTime.tm_sec = std::stoi(temp.substr(12, 2));
}

/**
* @brief Fills in the OS information for Linux systems
*/
void OperatingSystem::GetLux()
{
  std::string temp;
  std::time_t t = std::time(0);

  platform = "Linux";
  temp = runCommand("lsb_release -d");
  caption = trim(temp.substr(temp.find_first_of(":") + 1));
  serial = readFile("/sys/devices/virtual/dmi/id/product_serial", &temp) ? trim(temp) : "";
  temp = trim(runCommand("uname -m"));
  bit = endswith(temp, "64") || temp.find("armv8") != std::string::npos ? 64 : 32;
  temp = runCommand("lsb_release -r");
  version = new SemVer(trim(temp.substr(temp.find_first_of(":") + 1)), std::count(temp.begin(), temp.end(), '.') > 0 ? 0b11000u : 0b1000u);
  temp = runCommand("uname -r");
  kernel = new SemVer(trim(temp.substr(0, temp.find_first_of("-"))), 0b11000u);

  temp = runCommand("stat -c %y /var/cache/apt");
  installTime.tm_year = std::stoi(temp.substr(0, 4)) - 1900;
  installTime.tm_mon = std::stoi(temp.substr(5, 2)) - 1;
  installTime.tm_mday = std::stoi(temp.substr(8, 2));
  installTime.tm_hour = std::stoi(temp.substr(11, 2));
  installTime.tm_min = std::stoi(temp.substr(14, 2));
  installTime.tm_sec = std::stoi(temp.substr(17, 2));

  temp = runCommand("uptime -s");
  bootTime.tm_year = std::stoi(temp.substr(0, 4)) - 1900;
  bootTime.tm_mon = std::stoi(temp.substr(5, 2)) - 1;
  bootTime.tm_mday = std::stoi(temp.substr(8, 2));
  bootTime.tm_hour = std::stoi(temp.substr(11, 2));
  bootTime.tm_min = std::stoi(temp.substr(14, 2));
  bootTime.tm_sec = std::stoi(temp.substr(17, 2));

  curTime = (*std::localtime(&t));
}
#pragma endregion "Constructors' Assistants"

#pragma region "Operators"
/**
* @brief Reserves memory for a new OperatingSystem object
*
* @param  size  The amount of memory to allocate
* @return void* A pointer to the allocated memory
*/
void *OperatingSystem::operator new(std::size_t size)
{
  void *o = ::new (OperatingSystem);
  return o;
}

/**
* @brief Sets equal two OperatingSystem objects
*
* @param o The OperatingSystem object to copy from
*/
void OperatingSystem::operator=(const OperatingSystem &o)
{
  if (&o == this)
  {
    return;
  }
  platform = o.platform;
  caption = o.caption;
  serial = o.serial;
  bit = o.bit;
  installTime = o.installTime;
  bootTime = o.bootTime;
  curTime = o.curTime;
  kernel = o.kernel;
  version = o.version;
}

/**
* @brief Sets one OperatingSystem equal to another
*
* @param o The OperatingSystem to copy from
*/
void OperatingSystem::operator=(OperatingSystem *o)
{
  platform = o->platform;
  caption = o->caption;
  serial = o->serial;
  bit = o->bit;
  installTime = o->installTime;
  bootTime = o->bootTime;
  curTime = o->curTime;
  kernel = o->kernel;
  version = o->version;
}
#pragma endregion "Operators"

#pragma region "Static Methods"
/**
* @brief Checks if the given path is a file
*
* @param  path  The path to check
* @return true  The file at the path does exist
* @return false The file at the path does NOT exist
*/
bool fileExists(const std::string &path)
{
  if (std::FILE *file = std::fopen(path.c_str(), "r"))
  {
    std::fclose(file);
    return true;
  }
  else
  {
    return false;
  }
}

/**
* @brief Retrieves a given environment variable
*
* @param  key         The environment variable
* @return std::string The value of the given environment variable
*/
std::string getEnvVar(const std::string &key)
{
  char const *val = std::getenv(key.c_str());
  return val == NULL ? std::string() : std::string(val);
}

/**
* @brief Joins a list of paths with the system's path separator
*
* @param  paths       The list of paths to join
* @return std::string The joint path
*/
std::string joinPath(std::initializer_list<std::string> paths)
{
  std::stringstream buffer;
  for (std::string s : paths)
  {
    buffer << s;
    if (s != (*(paths.end() - 1)))
    {
      buffer << dirSep;
    }
  }
  return buffer.str();
}

/**
* @brief Gets the temporary directory for the system
*
* @return std::string The temporary directory
*/
std::string getTempDir()
{
  std::string tempDir;

  if (CGOGGLES_OS != OS_ERR)
  {
    if (tempDir.empty())
    {
      tempDir = getEnvVar("TMP");
    }
    if (tempDir.empty())
    {
      tempDir = getEnvVar("TEMP");
    }
    if (tempDir.empty())
    {
      tempDir = getEnvVar("TMPDIR");
    }
    if (tempDir.empty())
    {
      tempDir = getEnvVar("TEMPDIR");
    }
    if (tempDir.empty())
    {
      if (CGOGGLES_OS != OS_WIN)
      {
        tempDir = "/tmp";
      }
    }
  }

  return tempDir;
}

/**
* @brief Runs a command and retrieves output from stdout
*
* @param  cmd         The command to run
* @return std::string The command's output
*/
std::string runCommand(const std::string &cmd)
{
  std::array<char, 128> buffer;
  std::string result;
  std::unique_ptr<FILE, decltype(&P_PCLOSE)> pipe(P_POPEN(cmd.c_str(), "r"), P_PCLOSE);
  while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr)
  {
    result += buffer.data();
  }

  return result;
}

/**
* @brief Runs a WMIC query and retrieves the result
*
* @param  query       The query to run
* @param  path        The WMIC execution path
* @return std::string The query's result
*/
std::string runWmic(const std::string &query, std::string *path)
{
  std::string temp = runCommand((*path) + " " + query + " /format:list");
  std::stringstream buffer;

  for (size_t i = 0; i < temp.length(); ++i)
  {
    if (int(temp[i]) < 1 || int(temp[i]) > 0xFF)
    {
      continue;
    }
    buffer << temp[i];
  }
  temp = buffer.str();

  temp.erase(0, temp.find_first_of('=') + 1);
  temp.erase(temp.find_last_of('\r'));

  return temp;
}

/**
* @brief Runs a WMIC multi-key-query and retrieves the result
*
* @param  query                              The query to run
* @param  path                               The WMIC execution path
* @return std::map<std::string, std::string> The query's result
*/
std::map<std::string, std::string> runMultiWmic(const std::string &query, std::string *path)
{
  std::map<std::string, std::string> ret;
  std::string temp = runCommand((*path) + " " + query + " /format:list");
  std::stringstream buffer;
  std::vector<std::string> lines;

  for (size_t i = 0; i < temp.length(); ++i)
  {
    if (int(temp[i]) < 1 || int(temp[i]) > 0xFF)
    {
      continue;
    }
    buffer << temp[i];
  }
  temp = buffer.str();

  temp.erase(0, 4);
  temp.erase(temp.length() - 5);
  splitStringVector(temp, "\r\n", &lines);

  for (size_t i = 0; i < lines.size(); ++i)
  {
    ret[lines[i].substr(0, lines[i].find_first_of('='))] = lines[i].substr(lines[i].find_first_of('=') + 1);
  }

  return ret;
}

/**
* @brief Runs a WMIC multi-key-multi-object-query and retrieves the result
*
* @param query                                            The query to run
* @param path                                             The WMIC execution path
* @return std::vector<std::map<std::string, std::string>> The query's result
*/
std::vector<std::map<std::string, std::string>> runListMultiWmic(const std::string &query, std::string *path)
{
  std::vector<std::map<std::string, std::string>> ret;
  std::map<std::string, std::string> acc;
  std::string temp = runCommand((*path) + " " + query + " /format:list");
  std::stringstream buffer;
  std::vector<std::string> lines;

  for (size_t i = 0; i < temp.length(); ++i)
  {
    if (int(temp[i]) < 1 || int(temp[i]) > 0xFF)
    {
      continue;
    }
    buffer << temp[i];
  }
  temp = buffer.str();

  temp.erase(0, 4);
  temp.erase(temp.length() - 4);
  splitStringVector(temp, "\r\n", &lines);

  for (size_t i = 0; i < lines.size(); ++i)
  {
    if (i < lines.size() && lines[i] == "")
    {
      ret.push_back(acc);
      acc.clear();
      ++i;
      continue;
    }
    acc[lines[i].substr(0, lines[i].find_first_of('='))] = lines[i].substr(lines[i].find_first_of('=') + 1);
  }

  return ret;
}

/**
* @brief Get the WMIC path for Windows
*
* @return std::string The WMIC path
*/
std::string getWmicPath()
{
  std::string wmicPath = joinPath({getEnvVar("WINDIR"), "system32", "wbem", "wmic.exe"});

  return fileExists(wmicPath) ? wmicPath : "wmic.exe";
}
#pragma endregion "Static Methods"
